File: java/S5_DynamicBrowsing/src/sensors/FuelLevelSensor.java
/*
 *                         ADLINK Edge SDK
 *
 *   This software and documentation are Copyright 2018 to 2020 ADLINK
 *   Technology Limited, its affiliated companies and licensors. All rights
 *   reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 *
 */

/**
 * This code is part of example scenario 5 'Dynamic Browsing' of the
 * ADLINK Edge SDK. For a description of this scenario see the
 * 'Edge SDK User Guide' in the /doc directory of the Edge SDK instalation.
 *
 * For instructions on building and running the example see the README
 * file in the Edge SDK installation directory.
 */

package sensors;

import java.io.Closeable;
import java.util.Random;

import com.adlinktech.datariver.DataRiver;
import com.adlinktech.datariver.IotNvp;
import com.adlinktech.datariver.IotNvpSeq;
import com.adlinktech.datariver.IotValue;
import com.adlinktech.datariver.JSonTagGroupRegistry;
import com.adlinktech.datariver.JSonThingClassRegistry;
import com.adlinktech.datariver.JSonThingProperties;
import com.adlinktech.datariver.Thing;

public class FuelLevelSensor implements Closeable {
    private static final int FUEL_SAMPLE_DELAY_MS = 3000;
    private final String thingPropertiesUri;
    private final DataRiver dataRiver;
    private final Thing thing;
    
    public FuelLevelSensor(final String thingPropertiesUri) {
        System.out.println("Fuel Level Sensor started");
        this.thingPropertiesUri = thingPropertiesUri;
        this.dataRiver = DataRiver.getInstance();
        this.thing = createThing();
    }
    
    private void cleanup() {
        if(this.dataRiver != null) {
            this.dataRiver.close();
        }
        System.out.println("Fuel Level Sensor stopped");
    }
    
    private String getAbsFileUri(final String fileName) {
        final String dir = System.getProperty("user.dir");
        final String prefix = "file://";
        return prefix.concat(dir).concat("/").concat(fileName);
    }

    private Thing createThing() {
        // Create and Populate the TagGroup registry with JSON resource files.
        final JSonTagGroupRegistry tgr = new JSonTagGroupRegistry();
        tgr.registerTagGroupsFromUri(getAbsFileUri("definitions/TagGroup/com.adlinktech.example/SensorStateTagGroup.json"));
        tgr.registerTagGroupsFromUri(getAbsFileUri("definitions/TagGroup/com.adlinktech.example/FuelLevelTagGroup.json"));
        this.dataRiver.addTagGroupRegistry(tgr);

        // Create and Populate the ThingClass registry with JSON resource files.
        final JSonThingClassRegistry tcr = new JSonThingClassRegistry();
        tcr.registerThingClassesFromUri(getAbsFileUri("definitions/ThingClass/com.adlinktech.example/FuelLevelSensorThingClass.json"));
        this.dataRiver.addThingClassRegistry(tcr);

        // Create a Thing based on properties specified in a JSON resource file.
        final JSonThingProperties tp = new JSonThingProperties();
        tp.readPropertiesFromUri(this.thingPropertiesUri);
        return this.dataRiver.createThing(tp);
    }

    private void writeSample(final float level) {
        final IotValue levelValue = new IotValue();
        levelValue.setFloat32(level);
        final IotNvpSeq data = new IotNvpSeq();
        data.add(new IotNvp("level", levelValue));

        this.thing.write("level", data);
    }
    
    public void run(final int runningTime) {
        final Random rand = new Random();
        int sampleCount = (runningTime * 1000) / FUEL_SAMPLE_DELAY_MS;
        float fuelLevel = 1000.0f;

        while (sampleCount-- > 0) {
            // Simulate fuel level change
            fuelLevel -= (float)(rand.nextInt(100)) / 10.0f;
            writeSample(fuelLevel);

            try {
                Thread.sleep(FUEL_SAMPLE_DELAY_MS);
            } catch (InterruptedException e) {
                System.out.println("FuelLevelSensor is interrupted. Exiting.");
                break;
            }
        }
    }

    @Override
    public void close() {
        cleanup();
    }
}

File: dotnet/S5_DynamicBrowsing/Sensors/FuelLevelSensor.cs
/*                         ADLINK Edge SDK
 *
 *   This software and documentation are Copyright 2018 to 2020 ADLINK
 *   Technology Limited, its affiliated companies and licensors. All rights
 *   reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 *
 */

using System;
using Adlinktech.DataRiver;

namespace Sensors
{
    public class FuelLevelSensor : IDisposable
    {
        int SampleDelayMS = 3000;

        public bool Stop { get; set; }
        private DataRiver DataRiver { get; }
        private Thing Thing { get; }
        
        public FuelLevelSensor(string thingPropertiesURI)
        {
            DataRiver = DataRiver.GetInstance();            
            Thing = CreateThing(DataRiver, thingPropertiesURI);            

            Stop = false;
            Console.CancelKeyPress += this.CtrlCHandler;
            Console.WriteLine("Fuel Level Sensor started");
        }

        public void Dispose()
        {
            DataRiver.Close();
            DataRiver.Dispose();
            Console.WriteLine("Fuel Level Sensor stopped");
        }

        private void CtrlCHandler(object sender, ConsoleCancelEventArgs args)
        {
            Console.WriteLine("Trapped Ctrl-C. Stopping.");
            Stop = true;
            args.Cancel = true; // don't terminate the process, we'll do that.
        }

        private static Thing CreateThing(in DataRiver dataRiver, string thingPropertiesURI)
        {            
	    // Create and Populate the TagGroup registry with JSON resource files.
            var tgr = new JSonTagGroupRegistry();
            tgr.RegisterTagGroupsFromUri("file://./definitions/TagGroup/com.adlinktech.example/SensorStateTagGroup.json");
            tgr.RegisterTagGroupsFromUri("file://./definitions/TagGroup/com.adlinktech.example/FuelLevelTagGroup.json");
            dataRiver.AddTagGroupRegistry(tgr);

            // Create and Populate the ThingClass registry with JSON resource files.
            var tcr = new JSonThingClassRegistry();
            tcr.RegisterThingClassesFromUri("file://./definitions/ThingClass/com.adlinktech.example/FuelLevelSensorThingClass.json");
            dataRiver.AddThingClassRegistry(tcr);

            // Create a Thing based on properties specified in a JSON resource file.
            var tp = new JSonThingProperties();
            tp.ReadPropertiesFromUri(thingPropertiesURI);            

            return dataRiver.CreateThing(tp);
        }

        public int Run(ulong runningTime)
        {
            var rand = new Random();
            ulong sampleCount = (runningTime * 1000) / (ulong) SampleDelayMS; 
            float fuelLevel = 1000.0f;

            while (!Stop && (sampleCount-- > 0))
            {
                // Simulate fuel level change
                fuelLevel -= (float)(rand.Next() % 100) / 10.0f;
                WriteSample(fuelLevel);

                System.Threading.Thread.Sleep(SampleDelayMS);
            }             

            return 0;
        }  

        private void WriteSample(float level)
        {
            IotNvpSeq data = new IotNvpSeq();
            var levelValue = new IotValue();
            levelValue.Float32 = level;
            data.Add(new IotNvp("level", levelValue));

            Thing.Write("level", data);
        } 

        private void SetState(string state)
        {
            IotNvpSeq data = new IotNvpSeq();
            var stateValue = new IotValue();
            stateValue.String = state;
            data.Add(new IotNvp("state", stateValue));

            Thing.Write("state", data);
        }    

    }
}
